test/e2e/e2e_test.go

```diff
@@ -148,35 +151,109 @@ var _ = Describe("E2E", func() {
+					Expect(resources.DeleteResources(testEnv.Client, mgwFilePath)).To(Succeed())
 				})
 			})
 
-			It("sets up working ingress", func() {
-				if runtime.GOOS != "linux" {
-					Skip("MetalLB based test only works on Linux")
-				}
+			Context("when mgw deployment is created", func() {
+				var (
+					mgwNamespacedName types.NamespacedName
+					mgwDep            *appsv1.Deployment
+				)
+
+				BeforeEach(func() {
+					mgwNamespacedName = types.NamespacedName{
+						Namespace: testNamespace,
+						Name:      mgwName,
+					}
+				})
 
-				meshGatewayAddress, err := GetMeshGatewayAddress(testNamespace, "mgw01", 300*time.Second, 100*time.Millisecond)
-				Expect(err).NotTo(HaveOccurred())
+				JustBeforeEach(func() {
+					mgwDep, err = WaitForDeployment(istioTestEnv.c, mgwNamespacedName, 300*time.Second, 100*time.Second)
+					Expect(err).ShouldNot(HaveOccurred())
+				})
+
+				It("should only have one istio-proxy sidecar running", func() {
+					const (
+						mgwPodContainerAmount   int    = 1
+						istioProxyContainerName string = "istio-proxy"
+					)
+					// Validate if there one container only in mesh-gateway pod
+					containerList := GetContainersFromDeployment(mgwDep)
+					Expect(len(containerList)).Should(BeIdenticalTo(mgwPodContainerAmount))
+
+					// Check if the istio-proxy sidecar container exists in the pod
+					err = ContainerExists(containerList, istioProxyContainerName)
+					Expect(err).ShouldNot(HaveOccurred())
+				})
+
+			})
 
-				Expect(URLIsAccessible(log, fmt.Sprintf("http://%s:8080/get", meshGatewayAddress), 30*time.Second, 100*time.Millisecond)).
-					To(Succeed())
+			Context("when mgw service is created", func() {
+				var (
+					mgwNamespacedName        types.NamespacedName
+					mgwSvc                   *corev1.Service
+					mgwDeploymentMatchLabels map[string]string
+					mgwSvcLabelSelector      map[string]string
+				)
+
+				BeforeEach(func() {
+					mgwNamespacedName = types.NamespacedName{
+						Namespace: testNamespace,
+						Name:      mgwName,
+					}
+				})
+
+				JustBeforeEach(func() {
+					mgwDep, err := WaitForDeployment(istioTestEnv.c, mgwNamespacedName, 300*time.Second, 100*time.Second)
+					Expect(err).ShouldNot(HaveOccurred())
+					mgwDeploymentMatchLabels = mgwDep.Spec.Selector.MatchLabels
+
+					// Check if service is created that matches mesh-gateway's name in the same namespace
+					mgwSvc, err = GetService(context.TODO(), istioTestEnv.c, mgwNamespacedName)
+					Expect(err).ShouldNot(HaveOccurred())
+					mgwSvcLabelSelector = mgwSvc.Spec.Selector
+				})
+
+				It("should be pointed to the mesh gateway deployment", func() {
+					// Validate if mgw Service label selector matches mgw Deployment's MatchLabels field
+					Expect(mgwSvcLabelSelector).Should(BeEquivalentTo(mgwDeploymentMatchLabels))
+				})
+			})
+
+			Context("when mesh gateway address is created", func() {
+				// TODO: Add BeforeEach
```

Comment:
Please remove or fix this todo

Reply:
Thanks for the comment Peter. Just addressed it by removing `TODO`.

There is no really a practical use for us to add BeforeEach to there now.  
It is more of future usages when we could maybe add more test cases on that `"when mesh gateway address is created"` block.